### PR TITLE
Migrate to staert

### PR DESCRIPTION
**Pull request by cocap10**
Hi everybody,

Here a PR to migrate from viper/cobra to staert/flaeg, tow very new golang packages to easily load and merge program configuration structures.

[Flaeg](https://github.com/containous/flaeg) uses `StructTags` and `StructFields` names to build a complete CLI. It loads values from command line flags into the configuration structure thank's to reflection. It handles initialized configuration structure, default value on pointers and sub-commands. For more information, please visit https://github.com/containous/flaeg

[Staert](https://github.com/containous/staert) is another package which allows to merge configuration structures from several sources like a Toml file or the Flaeg package. It handles all [flaeg](https://github.com/containous/flaeg) features.

### CODE REVIEW

File: provider/boltdb.go [hunk 1]
```diff
@@ -9,7 +9,7 @@ import (
 
 // BoltDb holds configurations of the BoltDb provider.
 type BoltDb struct {
-	Kv `mapstructure:",squash"`
+	Kv `mapstructure:",squash" description:"go through"`
```
Comment (emilevauge) [inline provider/boltdb.go]: You can remove every `mapstructure:",squash"` now :)

**Comment by emilevauge** [comment]
Wating for https://github.com/containous/flaeg/issues/21
/cc @vdemeester can you review this one ?

File: configuration.go [hunk 2]
```diff
@@ -3,41 +3,44 @@ package main
...
 	"errors"
 	"fmt"
-	fmtlog "log"
-	"regexp"
-	"strings"
-	"time"
-
 	"github.com/containous/traefik/acme"
 	"github.com/containous/traefik/provider"
+	// fmtlog "log"
```
Comment (vdemeester) [inline configuration.go]: Not needed anymore :angel:

File: configuration.go [hunk 3]
```diff
@@ -46,6 +49,7 @@ type DefaultEntryPoints []string
 // String is the method to format the flag's value, part of the flag.Value interface.
 // The String method's output will be used in diagnostics.
 func (dep *DefaultEntryPoints) String() string {
+	//TODO :
```
Comment (vdemeester) [inline configuration.go]: What is left to do ? :angel: (should be written here, or the `TODO` should go away :stuck_out_tongue_closed_eyes: )

File: configuration.go [hunk 4]
```diff
@@ -63,9 +67,17 @@ func (dep *DefaultEntryPoints) Set(value string) error {
...
+func (dep *DefaultEntryPoints) Get() interface{} { return DefaultEntryPoints(*dep) }
+
+// SetValue sets the EntryPoints map with val
+func (dep *DefaultEntryPoints) SetValue(val interface{}) {
+	*dep = DefaultEntryPoints(val.(DefaultEntryPoints))
+}
+
 // Type is type of the struct
 func (dep *DefaultEntryPoints) Type() string {
-	return fmt.Sprint("defaultentrypoints²")
```
Comment (vdemeester) [inline configuration.go]: :joy_cat:

File: configuration.go [hunk 5]
```diff
@@ -74,6 +86,7 @@ type EntryPoints map[string]*EntryPoint
 // String is the method to format the flag's value, part of the flag.Value interface.
 // The String method's output will be used in diagnostics.
 func (ep *EntryPoints) String() string {
+	//TODO :
```
Comment (vdemeester) [inline configuration.go]: Same here :angel:

File: configuration.go [hunk 6]
```diff
@@ -190,117 +212,96 @@ type Retry struct {
...
-			Address: ":80",
-		}
-		configuration.DefaultEntryPoints = []string{"http"}
-	}
-
-	if configuration.File != nil && len(configuration.File.Filename) == 0 {
-		// no filename, setting to global config file
-		configuration.File.Filename = viper.ConfigFileUsed()
+// NewTraefikPointersConfiguration creates a TraefikConfiguration with pointers default values
+func NewTraefikPointersConfiguration() *TraefikConfiguration {
```
Comment (vdemeester) [inline configuration.go]: The name through me off a little bit :sweat:

File: integration/basic_test.go [hunk 7]
```diff
@@ -26,7 +26,7 @@ func (s *SimpleSuite) TestNoOrInexistentConfigShouldFail(c *check.C) {
 	time.Sleep(500 * time.Millisecond)
 	output := b.Bytes()
 
-	c.Assert(string(output), checker.Contains, "No configuration file found")
+	c.Assert(string(output), checker.Not(checker.Contains), "No configuration file found")
```
Comment (vdemeester) [inline integration/basic_test.go]: Hum so it should not print `No configuration file found` anymore. Do we still fail ? with another message ? or not ?
I feel like "if not", those tests are not needed anymore :angel:

File: provider/zk.go [hunk 8]
```diff
@@ -9,7 +9,7 @@ import (
 
 // Zookepper holds configurations of the Zookepper provider.
 type Zookepper struct {
-	Kv
+	Kv `description:"go through"`
```
Comment (vdemeester) [inline provider/zk.go]: Hum why is this here and not in the other `KV` based provider ? :stuck_out_tongue_closed_eyes:

File: traefik.go [hunk 9]
```diff
@@ -1,16 +1,130 @@
...
+	"strings"
 )
 
 func main() {
 	runtime.GOMAXPROCS(runtime.NumCPU())
-	if err := traefikCmd.Execute(); err != nil {
+
+	//traefik config inits
+	traefikConfiguration := NewTraefikConfiguration()
+	traefikPointersConfiguration := NewTraefikPointersConfiguration()
```
Comment (vdemeester) [inline traefik.go]: I need a little bit more *context* on that (it's a design decision from `flaeg` it seems, but I wonder why it is needed :stuck_out_tongue_closed_eyes:)

File: traefik.go [hunk 10]
```diff
@@ -1,16 +1,130 @@
...
+
+	//staert init
+	s := staert.NewStaert(traefikCmd)
+	//init toml source
+	toml := staert.NewTomlSource("traefik", []string{traefikConfiguration.ConfigFile, "/etc/traefik/", "$HOME/.traefik/", "."})
+
+	//add sources to staert
+	s.AddSource(toml)
+	s.AddSource(f)
+	if _, err := s.GetConfig(); err != nil {
```
Comment (vdemeester) [inline traefik.go]: I guess this populates something in the staert `s` but the name feels weird, `Get…` something makes me think I'm gonna get something in return, and there we don't care. I'm think it's just about validate the config, but I'm not sure :sweat_smile:

File: configuration.go (hunk 2 above)
Comment (cocap10) [inline configuration.go]: yep

File: provider/zk.go (hunk 8 above)
Comment (cocap10) [inline provider/zk.go]: We don't need that StructTag anymore !
I'll remove it. Thx ;)

File: configuration.go (hunk 6 above)
Comment (cocap10) [inline configuration.go]: Does it better with `NewTraefikDefaultPointersConfiguration` ?
This func returns a *TraefikConfiguration which contains default values under pointers fields (Like `Docker`,`File`,`Web`,`Marathon`,`Consul` ...)

Comment (emilevauge) [inline configuration.go]: I think he would prefer [`TemplateImporterRepositoryThreadFieldProxyGetterInterceptorGenericSerializer`](http://projects.haykranen.nl/java/) ^^

File: integration/basic_test.go (hunk 7 above)
Comment (cocap10) [inline integration/basic_test.go]: I'm not sure to understand, 
Staert doesn't fail if no configuration file found.
We should check nothing is printed in `output`, WDYT @emilevauge ?

File: traefik.go (hunk 9 above)
Comment (cocap10) [inline traefik.go]: `flaeg` loads values for args (or flags) into the configuration structure. 
Non pointers fields have can be initialized before `flaeg` process. Func `NewTraefikConfiguration` is in charge of this initialization. 
It's a bit trickier for pointers fields, their default value have to be stored in another structure: here it's `traefikPointersConfiguration`

File: integration/basic_test.go (hunk 7 above)
Comment (emilevauge) [inline integration/basic_test.go]: @vdemeester says we could maybe drop this test, and I agree with him :)

File: configuration.go (hunk 6 above)
Comment (cocap10) [inline configuration.go]: lol

Comment (vdemeester) [inline configuration.go]: @emilevauge :trollface: 
It feels weird to me to put *pointer* in a method name. Method names should not reflect what can of thing we manage (at least not at that level). Having two struct for the configuration is also a little bit weird :sweat_smile:

File: configuration.go (hunk 4 above)
Comment (emilevauge) [inline configuration.go]: This line is deleted ;)

File: traefik.go (hunk 10 above)
Comment (cocap10) [inline traefik.go]: I renamed the func from GetConfig to LoadConfig. WDYT ?
Have a look at this PR : https://github.com/containous/staert/pull/9#issuecomment-222076122

File: configuration.go [hunk 11]
```diff
@@ -47,6 +49,7 @@ type DefaultEntryPoints []string
 // String is the method to format the flag's value, part of the flag.Value interface.
 // The String method's output will be used in diagnostics.
 func (dep *DefaultEntryPoints) String() string {
+	//TODO : The string returned should be formatted in such way that the func Set below could parse it.
```
Comment (emilevauge) [inline configuration.go]: Should be implemented ;)

**Comment by emilevauge** [comment]
Great job @cocap10 :+1: 
LGTM!
/cc @vdemeester

**Comment by vdemeester** [comment]
LGTM 🐯
`ETOOMUCHCOMMITS` but I'll do :see_no_evil: for this one :joy_cat:
